master/internal/usergroup/api_groups.go

```diff
@@ -0,0 +1,213 @@
+		req.Name, model.UserID(req.UserId), int(req.Offset), int(req.Limit))
+	if err != nil {
+		return nil, err
+	}
+
+	searchResults := make([]*groupv1.GroupSearchResult, len(groups))
+	for i, g := range groups {
+		searchResults[i] = &groupv1.GroupSearchResult{
+			Group:      g.Proto(),
+			NumMembers: memberCounts[i],
+		}
+	}
+
+	return &apiv1.GetGroupsResponse{
+		Groups: searchResults,
+		Pagination: &apiv1.Pagination{
+			Offset:     req.Offset,
+			Limit:      req.Limit,
+			StartIndex: req.Offset,
+			EndIndex:   req.Offset + int32(len(groups)),
+			Total:      int32(tableCount),
+		},
+	}, nil
+}
+
+// GetGroup finds and returns details of the group specified.
+func (a *APIServer) GetGroup(ctx context.Context, req *apiv1.GetGroupRequest,
+) (resp *apiv1.GetGroupResponse, err error) {
+	// Detect whether we're returning special errors and convert to gRPC error
+	defer func() {
+		err = mapAndFilterErrors(err)
+	}()
+
+	gid := int(req.GroupId)
+	g, err := GroupByIDTx(ctx, nil, gid)
+	if err != nil {
+		return nil, err
+	}
+
+	users, err := UsersInGroupTx(ctx, nil, gid)
+	if err != nil {
+		return nil, err
+	}
+
+	gDetail := groupv1.GroupDetails{
+		GroupId: int32(g.ID),
+		Name:    g.Name,
+		Users:   model.Users(users).Proto(),
+	}
+
+	return &apiv1.GetGroupResponse{
+		Group: &gDetail,
+	}, nil
+}
+
+// UpdateGroup updates the group and returns the newly updated group details.
+func (a *APIServer) UpdateGroup(ctx context.Context, req *apiv1.UpdateGroupRequest,
+) (resp *apiv1.UpdateGroupResponse, err error) {
+	// Detect whether we're returning special errors and convert to gRPC error
+	defer func() {
+		err = mapAndFilterErrors(err)
+	}()
+
+	var addUsers []model.UserID
+	var removeUsers []model.UserID
+
+	if len(req.AddUsers) > 0 {
+		addUsers = intsToUserIDs(req.AddUsers)
+	}
+
+	if len(req.RemoveUsers) > 0 {
+		removeUsers = intsToUserIDs(req.RemoveUsers)
+	}
+
+	users, newName, err := UpdateGroupAndMembers(ctx,
+		int(req.GroupId), req.Name, addUsers, removeUsers)
+	if err != nil {
+		return nil, err
+	}
+
+	resp = &apiv1.UpdateGroupResponse{
+		Group: &groupv1.GroupDetails{
+			GroupId: req.GroupId,
+			Name:    newName,
+			Users:   model.Users(users).Proto(),
+		},
+	}
+
+	return resp, nil
+}
+
+// DeleteGroup deletes the database entry for the group.
+func (a *APIServer) DeleteGroup(ctx context.Context, req *apiv1.DeleteGroupRequest,
+) (resp *apiv1.DeleteGroupResponse, err error) {
+	// Detect whether we're returning special errors and convert to gRPC error
+	defer func() {
+		err = mapAndFilterErrors(err)
+	}()
+
+	err = DeleteGroup(ctx, int(req.GroupId))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.DeleteGroupResponse{}, nil
+}
+
+func intsToUserIDs(ints []int32) []model.UserID {
+	ids := make([]model.UserID, len(ints))
+
+	for i := range ints {
+		ids[i] = model.UserID(ints[i])
+	}
+
+	return ids
+}
+
+const (
+	maxLimit = 500
+)
+
+var (
+	errBadRequest   = status.Error(codes.InvalidArgument, "Bad request")
```

Comment:
nit: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
Hm, yeah, that's interesting. Since this is what's sent to HTTP clients, the goal was to mimic the standard library in places like [this](https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/net/http/server.go;l=1931), but it seems like I picked an awkward not-quite-consistent middle ground. I'll go with lowercase.